File: src/ui/theme.js
import { existsSync, readFileSync } from 'node:fs';
import { homedir } from 'node:os';
import { join } from 'node:path';

/**
 * Parses simple key = value pairs from TOML color files (Omarchy / Linux themes)
 */
function parseTomlColors(filePath) {
  if (!existsSync(filePath)) return null;
  try {
    const raw = readFileSync(filePath, 'utf8');
    const colors = {};
    for (const line of raw.split('\n')) {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith('#') || !trimmed.includes('=')) continue;
      const [k, v] = trimmed.split('=', 1).map((s) => s.trim());
      const rest = trimmed.slice(trimmed.indexOf('=') + 1).trim().replace(/^["']|["']$/g, '');
      if (rest.startsWith('#') || rest.startsWith('rgb') || rest.startsWith('rgba')) {
        colors[k] = rest;
      }
    }
    return colors;
  } catch {
    return null;
  }
}

/**
 * Detects current Linux desktop / Omarchy theme colors
 */
export function detectSystemTheme() {
  const home = homedir();
  const omarchyStatePath = join(home, '.local', 'state', 'omarchy', 'current', 'theme', 'colors.toml');
  const omarchyConfigPath = join(home, '.config', 'omarchy', 'current', 'theme', 'colors.toml');
  const pywalCachePath = join(home, '.cache', 'wal', 'colors.json');

  // 1. Check Omarchy Dynamic System Theme
  const omarchyColors = parseTomlColors(omarchyStatePath) || parseTomlColors(omarchyConfigPath);
  if (omarchyColors) {
    const accent = omarchyColors.accent || omarchyColors.active_border_color || omarchyColors.cyan || '#7aa2f7';
    const bgDark = omarchyColors.darker_background || omarchyColors.dark_background || omarchyColors.background || '#101315';
    const bgPanel = omarchyColors.background || omarchyColors.dark_background || '#15181b';
    const fg = omarchyColors.foreground || omarchyColors.bright_foreground || '#cacccc';
    const muted = omarchyColors.muted || omarchyColors.dark_foreground || '#4b4e55';
    const borderDim = omarchyColors.selection || omarchyColors.muted || '#343d41';
    const borderFocus = omarchyColors.active_border_color || accent;

    return {
      id: 'SYSTEM_AUTO',
      name: '🖥️ System Auto-Sync (Omarchy / OS Theme)',
      bgDark,
      bgPanel,
      bgLcd: omarchyColors.darker_background || bgDark,
      borderDim,
      borderFocus,
      borderLcd: borderDim,
      amber: accent,
      amberBright: omarchyColors.light_foreground || omarchyColors.bright_yellow || fg,
      amberDim: omarchyColors.blue || omarchyColors.muted || muted,
      greenPhosphor: omarchyColors.green || omarchyColors.bright_green || '#9fa5a9',
      greenDim: borderDim,
      gold: omarchyColors.yellow || omarchyColors.bright_yellow || '#d9dbdc',
      cyanDolby: omarchyColors.cyan || omarchyColors.bright_cyan || accent,
      redLed: omarchyColors.bright_red || omarchyColors.red || '#de6145',
      yellowLed: omarchyColors.yellow || '#d9dbdc',
      cream: fg,
      chrome: omarchyColors.bright_foreground || fg,
      muted
    };
  }

  // 2. Check Pywal System Colors
  if (existsSync(pywalCachePath)) {
    try {
      const walData = JSON.parse(readFileSync(pywalCachePath, 'utf8'));
      if (walData?.colors) {
        const c = walData.colors;
        const special = walData.special || {};
        return {
          id: 'SYSTEM_AUTO',
          name: '🖥️ System Auto-Sync (Pywal / Wal Theme)',
          bgDark: special.background || c.color0 || '#11111b',
          bgPanel: c.color0 || '#181825',
          bgLcd: special.background || '#11111b',
          borderDim: c.color8 || '#45475a',
          borderFocus: c.color4 || c.color6 || '#cba6f7',
          borderLcd: c.color8 || '#45475a',
          amber: c.color3 || c.color4 || '#fabd2f',
          amberBright: c.color11 || c.color15 || '#ffd24d',
          amberDim: c.color8 || '#89b4fa',
          greenPhosphor: c.color2 || c.color10 || '#a6e3a1',
          greenDim: c.color8 || '#313244',
          gold: c.color3 || c.color11 || '#f9e2af',
          cyanDolby: c.color6 || c.color14 || '#89dceb',
          redLed: c.color1 || c.color9 || '#f38ba8',
          yellowLed: c.color3 || '#fab387',
          cream: special.foreground || c.color7 || c.color15 || '#cdd6f4',
          chrome: c.color7 || '#bac2de',
          muted: c.color8 || '#6c7086'
        };
      }
    } catch {}
  }

  // Fallback to Amber Gold
  return null;
}

export const THEMES = {
  SYSTEM_AUTO: {
    id: 'SYSTEM_AUTO',
    name: '🖥️ System Auto-Sync (Omarchy / OS Theme)',
    bgDark: '#101315',
    bgPanel: '#15181b',
    bgLcd: '#0c0e10',
    borderDim: '#343d41',
    borderFocus: '#798186',
    borderLcd: '#22282c',
    amber: '#798186',
    amberBright: '#cacccc',
    amberDim: '#4b4e55',
    greenPhosphor: '#9fa5a9',
    greenDim: '#22282c',
    gold: '#d9dbdc',
    cyanDolby: '#707070',
    redLed: '#de6145',
    yellowLed: '#d9dbdc',
    cream: '#cacccc',
    chrome: '#a5aeb4',
    muted: '#4b4e55'
  },
  AMBER_GOLD: {
    id: 'AMBER_GOLD',
    name: '📻 Vintage Amber Gold (Nakamichi Hi-Fi)',
    bgDark: '#0a0d13',
    bgPanel: '#11151f',
    bgLcd: '#06130a',
    borderDim: '#243348',
    borderFocus: '#ffb000',
    borderLcd: '#1b4d24',
    amber: '#ffb000',
    amberBright: '#ffd24d',
    amberDim: '#996900',
    greenPhosphor: '#33ff33',
    greenDim: '#114a1a',
    gold: '#f5c542',
    cyanDolby: '#00e5ff',
    redLed: '#ff3344',
    yellowLed: '#ffee33',
    cream: '#f3ead8',
    chrome: '#b8c4ce',
    muted: '#6f7e91'
  },
  CATPPUCCIN_MOCHA: {
    id: 'CATPPUCCIN_MOCHA',
    name: '☕ Catppuccin Mocha (Pastel Lavender)',
    bgDark: '#11111b',
    bgPanel: '#181825',
    bgLcd: '#1e1e2e',
    borderDim: '#45475a',
    borderFocus: '#cba6f7',
    borderLcd: '#585b70',
    amber: '#cba6f7', // Mauve
    amberBright: '#f5c2e7', // Pink
    amberDim: '#89b4fa', // Blue
    greenPhosphor: '#a6e3a1', // Green
    greenDim: '#313244',
    gold: '#f9e2af', // Yellow
    cyanDolby: '#89dceb', // Sky
    redLed: '#f38ba8', // Red
    yellowLed: '#fab387', // Peach
    cream: '#cdd6f4', // Text
    chrome: '#bac2de', // Subtext
    muted: '#6c7086' // Overlay
  },
  TOKYO_NIGHT: {
    id: 'TOKYO_NIGHT',
    name: '🌃 Tokyo Night (Neon Storm)',
    bgDark: '#16161e',
    bgPanel: '#1a1b26',
    bgLcd: '#13141c',
    borderDim: '#292e42',
    borderFocus: '#7aa2f7',
    borderLcd: '#414868',
    amber: '#7aa2f7', // Blue
    amberBright: '#7dcfff', // Cyan
    amberDim: '#3d59a1',
    greenPhosphor: '#73daca', // Teal Green
    greenDim: '#1f2335',
    gold: '#e0af68', // Yellow
    cyanDolby: '#2ac3de',
    redLed: '#f7768e', // Red
    yellowLed: '#ff9e64', // Orange
    cream: '#c0caf5',
    chrome: '#a9b1d6',
    muted: '#565f89'
  },
  GRUVBOX_RETRO: {
    id: 'GRUVBOX_RETRO',
    name: '🍂 Gruvbox Retro (Warm Analog)',
    bgDark: '#1d2021',
    bgPanel: '#282828',
    bgLcd: '#141617',
    borderDim: '#504945',
    borderFocus: '#fe8019',
    borderLcd: '#3c3836',
    amber: '#fe8019', // Orange
    amberBright: '#fabd2f', // Yellow
    amberDim: '#af3a03',
    greenPhosphor: '#b8bb26', // Green
    greenDim: '#32302f',
    gold: '#fabd2f',
    cyanDolby: '#8ec07c', // Aqua
    redLed: '#fb4934',
    yellowLed: '#d79921',
    cream: '#ebdbb2',
    chrome: '#d5c4a1',
    muted: '#928374'
  },
  NORD_FROST: {
    id: 'NORD_FROST',
    name: '❄️ Nord Frost (Arctic Ice)',
    bgDark: '#242933',
    bgPanel: '#2e3440',
    bgLcd: '#1e222a',
    borderDim: '#434c5e',
    borderFocus: '#88c0d0',
    borderLcd: '#3b4252',
    amber: '#88c0d0', // Frost Cyan
    amberBright: '#8fbcbb',
    amberDim: '#5e81ac',
    greenPhosphor: '#a3be8c', // Nord Green
    greenDim: '#2e3440',
    gold: '#ebcb8b', // Nord Yellow
    cyanDolby: '#81a1c1',
    redLed: '#bf616a',
    yellowLed: '#d08770',
    cream: '#eceff4',
    chrome: '#e5e9f0',
    muted: '#4c566a'
  },
  DRACULA: {
    id: 'DRACULA',
    name: '🧛 Dracula (Vampire Night)',
    bgDark: '#1e1f29',
    bgPanel: '#282a36',
    bgLcd: '#191a21',
    borderDim: '#44475a',
    borderFocus: '#bd93f9',
    borderLcd: '#6272a4',
    amber: '#bd93f9', // Purple
    amberBright: '#ff79c6', // Pink
    amberDim: '#6272a4',
    greenPhosphor: '#50fa7b', // Green
    greenDim: '#282a36',
    gold: '#f1fa8c', // Yellow
    cyanDolby: '#8be9fd', // Cyan
    redLed: '#ff5555',
    yellowLed: '#ffb86c', // Orange
    cream: '#f8f8f2',
    chrome: '#e6e6e6',
    muted: '#6272a4'
  },
  MATRIX_GREEN: {
    id: 'MATRIX_GREEN',
    name: '📟 Matrix Cyber Green (Hacker CRT)',
    bgDark: '#030a04',
    bgPanel: '#071509',
    bgLcd: '#020603',
    borderDim: '#0d3814',
    borderFocus: '#00ff41',
    borderLcd: '#0b2e10',
    amber: '#00ff41', // Matrix Green
    amberBright: '#33ff66',
    amberDim: '#008f11',
    greenPhosphor: '#00ff41',
    greenDim: '#003b00',
    gold: '#88ff88',
    cyanDolby: '#00e5ff',
    redLed: '#ff3344',
    yellowLed: '#ffff33',
    cream: '#e0ffe4',
    chrome: '#a0e6a8',
    muted: '#286830'
  }
};

export const THEME_LIST = Object.values(THEMES);
export const THEME_KEYS = Object.keys(THEMES);

export function getTheme(themeId) {
  if (themeId === 'SYSTEM_AUTO' || !themeId) {
    const detected = detectSystemTheme();
    if (detected) return detected;
    return THEMES.SYSTEM_AUTO;
  }
  return THEMES[themeId] || THEMES.AMBER_GOLD;
}

export function cycleTheme(currentId, delta = 1) {
  const idx = THEME_KEYS.indexOf(currentId);
  const nextIdx = (idx + delta + THEME_KEYS.length) % THEME_KEYS.length;
  return THEME_KEYS[nextIdx];
}

File: src/ui/theme.rs
use ratatui::style::Color;
use std::fs;

#[derive(Debug, Clone)]
pub struct Theme {
    pub id: &'static str,
    pub name: &'static str,
    pub bg_dark: Color,
    pub bg_panel: Color,
    pub bg_lcd: Color,
    pub border_dim: Color,
    pub border_focus: Color,
    pub border_lcd: Color,
    pub amber: Color,
    pub amber_bright: Color,
    pub green_phosphor: Color,
    pub gold: Color,
    pub cyan_dolby: Color,
    pub red_led: Color,
    pub yellow_led: Color,
    pub cream: Color,
    pub chrome: Color,
    pub muted: Color,
}

pub fn hex(h: &str) -> Color {
    let clean = h.trim_start_matches('#');
    if clean.len() == 6 {
        let r = u8::from_str_radix(&clean[0..2], 16).unwrap_or(0);
        let g = u8::from_str_radix(&clean[2..4], 16).unwrap_or(0);
        let b = u8::from_str_radix(&clean[4..6], 16).unwrap_or(0);
        Color::Rgb(r, g, b)
    } else {
        Color::Reset
    }
}

pub fn detect_system_theme() -> Option<Theme> {
    let home = dirs::home_dir()?;
    let paths = [
        home.join(".local/state/omarchy/current/theme/colors.toml"),
        home.join(".config/omarchy/current/theme/colors.toml"),
    ];

    for path in paths {
        if path.exists() {
            if let Ok(content) = fs::read_to_string(path) {
                return parse_omarchy_toml(&content);
            }
        }
    }
    None
}

fn parse_omarchy_toml(content: &str) -> Option<Theme> {
    let val: toml::Value = toml::from_str(content).ok()?;
    let get_color = |k: &str| -> Option<Color> {
        val.get(k).and_then(|v| v.as_str()).map(hex)
    };

    let accent = get_color("accent").or_else(|| get_color("cyan")).unwrap_or(hex("#ffb000"));
    let bg = get_color("darker_background").or_else(|| get_color("background")).unwrap_or(hex("#101315"));
    let panel = get_color("background").unwrap_or(hex("#15181b"));
    let fg = get_color("foreground").unwrap_or(hex("#cacccc"));
    let muted = get_color("muted").unwrap_or(hex("#4b4e55"));

    Some(Theme {
        id: "SYSTEM_AUTO",
        name: "🖥️ Omarchy Auto",
        bg_dark: bg,
        bg_panel: panel,
        bg_lcd: bg,
        border_dim: get_color("selection").unwrap_or(hex("#343d41")),
        border_focus: accent,
        border_lcd: get_color("selection").unwrap_or(hex("#343d41")),
        amber: accent,
        amber_bright: get_color("bright_yellow").unwrap_or(fg),
        green_phosphor: get_color("green").unwrap_or(hex("#9fa5a9")),
        gold: get_color("yellow").unwrap_or(hex("#eed49f")),
        cyan_dolby: get_color("cyan").unwrap_or(hex("#8bd5ca")),
        red_led: get_color("red").unwrap_or(hex("#ed8796")),
        yellow_led: get_color("yellow").unwrap_or(hex("#eed49f")),
        cream: fg,
        chrome: get_color("light_foreground").unwrap_or(fg),
        muted,
    })
}

pub fn get_themes() -> Vec<Theme> {
    let mut list = Vec::new();
    if let Some(sys) = detect_system_theme() {
        list.push(sys);
    }
    list.extend([
        Theme {
            id: "AMBER_GOLD",
            name: "📼 Amber Gold",
            bg_dark: hex("#0e0b06"),
            bg_panel: hex("#16120a"),
            bg_lcd: hex("#1f190e"),
            border_dim: hex("#3a2f1c"),
            border_focus: hex("#ffb000"),
            border_lcd: hex("#3a2f1c"),
            amber: hex("#ffb000"),
            amber_bright: hex("#ffd24d"),
            green_phosphor: hex("#33ff33"),
            gold: hex("#f5c542"),
            cyan_dolby: hex("#00e5ff"),
            red_led: hex("#ff3344"),
            yellow_led: hex("#ffff33"),
            cream: hex("#f3ead8"),
            chrome: hex("#b8c4ce"),
            muted: hex("#6f7e91"),
        },
        Theme {
            id: "CYBER_NEON",
            name: "⚡ Cyber Neon",
            bg_dark: hex("#080512"),
            bg_panel: hex("#0f0c20"),
            bg_lcd: hex("#130f2c"),
            border_dim: hex("#2c2250"),
            border_focus: hex("#00f0ff"),
            border_lcd: hex("#332460"),
            amber: hex("#ff007f"),
            amber_bright: hex("#ff55a3"),
            green_phosphor: hex("#00ff9f"),
            gold: hex("#ffe600"),
            cyan_dolby: hex("#00f0ff"),
            red_led: hex("#ff0055"),
            yellow_led: hex("#ffff00"),
            cream: hex("#e0f7fa"),
            chrome: hex("#a5b4fc"),
            muted: hex("#625f87"),
        },
        Theme {
            id: "GREEN_PHOSPHOR",
            name: "📟 Green Phosphor",
            bg_dark: hex("#040d06"),
            bg_panel: hex("#08180c"),
            bg_lcd: hex("#0c2010"),
            border_dim: hex("#1a3b22"),
            border_focus: hex("#00ff66"),
            border_lcd: hex("#1d4a27"),
            amber: hex("#00ff66"),
            amber_bright: hex("#66ff99"),
            green_phosphor: hex("#00ff66"),
            gold: hex("#a3e635"),
            cyan_dolby: hex("#2dd4bf"),
            red_led: hex("#f87171"),
            yellow_led: hex("#facc15"),
            cream: hex("#dcfce7"),
            chrome: hex("#86efac"),
            muted: hex("#3b6b47"),
        },
        Theme {
            id: "TOKYO_NIGHT",
            name: "🌃 Tokyo Night",
            bg_dark: hex("#16161e"),
            bg_panel: hex("#1a1b26"),
            bg_lcd: hex("#24283b"),
            border_dim: hex("#3b4261"),
            border_focus: hex("#7aa2f7"),
            border_lcd: hex("#414868"),
            amber: hex("#7aa2f7"),
            amber_bright: hex("#bb9af7"),
            green_phosphor: hex("#73daca"),
            gold: hex("#e0af68"),
            cyan_dolby: hex("#7dcfff"),
            red_led: hex("#f7768e"),
            yellow_led: hex("#e0af68"),
            cream: hex("#c0caf5"),
            chrome: hex("#9aa5ce"),
            muted: hex("#565f89"),
        },
        Theme {
            id: "CATPPUCCIN",
            name: "☕ Catppuccin Mocha",
            bg_dark: hex("#11111b"),
            bg_panel: hex("#181825"),
            bg_lcd: hex("#1e1e2e"),
            border_dim: hex("#45475a"),
            border_focus: hex("#f5c2e7"),
            border_lcd: hex("#585b70"),
            amber: hex("#fab387"),
            amber_bright: hex("#f9e2af"),
            green_phosphor: hex("#a6e3a1"),
            gold: hex("#f9e2af"),
            cyan_dolby: hex("#89dceb"),
            red_led: hex("#f38ba8"),
            yellow_led: hex("#f9e2af"),
            cream: hex("#cdd6f4"),
            chrome: hex("#bac2de"),
            muted: hex("#6c7086"),
        },
        Theme {
            id: "NORD",
            name: "❄️ Nord Frost",
            bg_dark: hex("#242933"),
            bg_panel: hex("#2e3440"),
            bg_lcd: hex("#3b4252"),
            border_dim: hex("#4c566a"),
            border_focus: hex("#88c0d0"),
            border_lcd: hex("#434c5e"),
            amber: hex("#88c0d0"),
            amber_bright: hex("#8fbcbb"),
            green_phosphor: hex("#a3be8c"),
            gold: hex("#ebcb8b"),
            cyan_dolby: hex("#81a1c1"),
            red_led: hex("#bf616a"),
            yellow_led: hex("#ebcb8b"),
            cream: hex("#eceff4"),
            chrome: hex("#d8dee9"),
            muted: hex("#616e88"),
        },
    ]);
    list
}
